contracts/staking/StakingPool.sol

```diff
@@ -0,0 +1,470 @@
+        uint128[5] rewardAmounts; // should we make this a set so we can check for uniqueness?
+    }
+
+    mapping(address => User) private _users;
+
+    StakingPoolLib.Data private _stakingPoolInfo;
+
+    event WithdrawRewards(
+        address indexed user,
+        address rewardToken,
+        uint256 rewards
+    );
+
+    event WithdrawStake(address indexed user, uint256 stake);
+
+    event Deposit(address indexed user, uint256 depositAmount);
+
+    event InitializeRewards(address rewardTokens, uint256 amount);
+
+    event RewardsAvailableTimestamp(uint32 rewardsAvailableTimestamp);
+
+    modifier rewardsAvailable() {
+        require(_isRewardsAvailable(), "StakingPool: rewards too early");
+        _;
+    }
+
+    modifier stakingPeriodComplete() {
+        require(_isStakingPeriodComplete(), "StakingPool: still stake period");
+        _;
+    }
+
+    modifier stakingPoolRequirementsUnmet() {
+        //slither-disable-next-line timestamp
+        require(
+            (_stakingPoolInfo.totalStakedAmount <
+                _stakingPoolInfo.minTotalPoolStake) &&
+                (block.timestamp > _stakingPoolInfo.epochStartTimestamp),
+            "StakingPool: requirements unmet"
+        );
+        _;
+    }
+
+    modifier emergencyModeEnabled() {
+        require(
+            _stakingPoolInfo.emergencyMode,
+            "StakingPool: not emergency mode"
+        );
+        _;
+    }
+
+    modifier stakingPeriodNotStarted() {
+        //slither-disable-next-line timestamp
+        require(
+            block.timestamp >= _stakingPoolInfo.epochStartTimestamp,
+            "StakingPool: too early"
+        );
+        _;
+    }
+
+    modifier stakingPoolNotFull(uint256 _deposit) {
+        require(
+            _stakingPoolInfo.totalStakedAmount + _deposit <
+                _stakingPoolInfo.maxTotalPoolStake,
+            "StakingPool: pool full"
+        );
+        _;
+    }
+
+    function pause()
+        external
+        whenNotPaused
+        atLeastDaoMeepleRole(_stakingPoolInfo.daoId)
+    {
+        _pause();
+    }
+
+    function unpause()
+        external
+        whenPaused
+        atLeastDaoMeepleRole(_stakingPoolInfo.daoId)
+    {
+        _unpause();
+    }
+
+    /**
+     * @notice Only entry point for a user to deposit into the staking pool
+     *
+     * @param amount Amount of stake tokens to deposit
+     */
+    function deposit(uint256 amount)
+        external
+        whenNotPaused
+        stakingPeriodNotStarted
+        nonReentrant
+        stakingPoolNotFull(amount)
+    {
+        require(
+            amount >= _stakingPoolInfo.minimumContribution,
+            "StakingPool: min contribution"
+        );
+
+        User storage user = _users[_msgSender()];
+        StakingPoolLib.Data storage _info = _stakingPoolInfo;
+
+        user.depositAmount += uint128(amount);
+        _info.totalStakedAmount += uint128(amount);
+
+        // calculate/update rewards
```

Comment:
suggestion: refactor into private helper function and remove the comments.

When you feel the need or value add of an inline comment, consider moving the following code into a private helper - clean code 😁

Reply:
so I created `function _calculateFixedRewards(StakingPoolLib.Data memory _info, User storage user, uint256 amount) private` 

and

`function _updateRewardsRatios(StakingPoolLib.Data storage _info) private` 

but now the contract size is too large..

Reply:
Focus on getting this into the main branch, then we can trim the size down.